provisioning_templates/snippet/pxegrub2_chainload.erb

```diff
@@ -5,22 +5,37 @@ model: ProvisioningTemplate
 snippet: true
 %>
 <%
-  paths = ["fedora", "redhat", "centos", "debian", "ubuntu", "sles", "opensuse", "Microsoft", "EFI"]
+  efi_file_list = ["shim.efi", "grubx64.efi", "bootmgfw.efi" ]
 -%>
 menuentry 'Chainload Grub2 EFI from ESP' --id local {
-  unset root
-  echo Chainloading Grub2 EFI from ESP, available devices:
-  ls
-  echo -n "Probing ESP partition ... "
-  search --file --no-floppy --set=root /EFI/BOOT/BOOTX64.EFI
-  echo found $root
-  sleep 2
-  if [ -f ($root)/EFI/BOOT/grubx64.efi ]; then
-    chainloader ($root)/EFI/BOOT/grubx64.efi
-  <% paths.each do |path| %>
-  elif [ -f ($root)/EFI/<%= path %>/grubx64.efi ]; then
-    chainloader ($root)/EFI/<%= path %>/grubx64.efi
+  insmod part_gpt
+  insmod fat
+  insmod chain
+
+  <% efi_file_list.each do |search_efi| %>
+  if [ $efi_found != true ]; then
+    echo "Search system for <%= search_efi %>"
+    for efi in (hd*,gpt*)/efi/*/*.efi (hd*,gpt*)/efi/*/*/*.efi (hd*,gpt*)/*.efi (hd*,gpt*)/*/*.efi ; do
+      regexp --set=1:efi_device '^\((.*)\)/' "${efi}"
+      regexp --set=1:efi_path '(.*)/.*$' "${efi}"
+      regexp --set=1:efi_file '.*\/(.*)$' "${efi}"
```

Comment:
Can't we use `search` command for that? This is a bit hacky, I am wondering why `search` is not used.

Reply:
the search command does not accept wildcards, one has to specify the exact location of the efi file you're searching. It would require many statements (at least one for each distribution) to determine the root disk. Although  I agree that the for loop is hacky (and I'm not a programmer.) indeed, it was one way I could get it working.

Reply:
As for the slowness, the for loop should have eliminated the fd0 and cd0 searches. Unfortunately, it simply ignores this.. Do you have any idea why?